Create /listen overlay endpoint

louisheal opened the pull request.

louisheal committed:
create overlay repository, service, controller and routes

louisheal committed:
create listen endpoint
write overlay controller tests

louisheal committed:
added tests for authentication service

louisheal committed:
write channel service tests

louisheal committed:
write twitch repository tests

louisheal committed:
move channel tests to correct file

louisheal committed:
write twitch api tests

louisheal commented on controllers/overlay.go:
```diff
@@ -0,0 +1,82 @@
+	})
+}
+
+func (c *OverlayController) HandleViewers(ctx *gin.Context) {
```
Remove for this PR

louisheal commented on main.go:
```diff
@@ -13,11 +13,17 @@ func main() {
+	}
+
+	// This smells
+	twitchRepo := config.CreateTwitchRepo()
```
Need some guidance with how to idiomatically setup the Twitch "Repo"

kenjdavidson commented on controllers/overlay.go:
```diff
@@ -0,0 +1,82 @@
+	OverlayID uuid.UUID
+}
+
+type channelService interface {
```
I feel like there must be some **Go** advanced reasoning for doing this, which I'm super curious about since this is the first time I've actually looked at Go, it still looks odd having these two interfaces here solely for typing the `OverlayController`.

louisheal replied:
Definitely an anti pattern in other languages but in Go you "accept interfaces and return structs"

kenjdavidson replied:
Oh, I get the accept interfaces and return concrete thing, my comment was more on why they are not defined in the same package I guess.  I was more curious about the, I guess, private (lower case) nature of `channelService` in here, is that a pattern? Or is the pattern to try and redesign the packages so that you have both interface and implementation available.  For example, you may have another controller that may need the channel service, in which case this will need to be duplicated in there.

kenjdavidson commented on models/channel.go:
```diff
@@ -3,7 +3,7 @@ package models
 
 type Channel struct {
-	ChannelID   string
+	ChannelID   UserID
```
Should this be `ChannelId` defined in the same manner as `UserId`?  If they will always be the same (ie. string, uuid, etc) then maybe just a `ModelId` or `Id` type?

louisheal replied:
ChannelID and UserID are both Twitch user ids
Should they be separate types, or will renaming to just ID make more sense?

kenjdavidson commented on repositories/twitch.go:
```diff
@@ -0,0 +1,93 @@
+		return "", err
+	}
+
+	accessToken, err := repo.getAccessToken()
```
The app access token should probably be retrieved on creation or within a separate call after creation.  Presumably you're going to have more methods than `GetUsername` and the app token stays valid for a longer period of time than user tokens.   Don't think you want to re-call it each time someone calls `GetUsername`.  Not sure what the `Go` way of doing this would be, but normally you would either setup a timer based on the expiry date, or automatically refresh on the first 401 and try again.

kenjdavidson commented on services/channel.go:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	stream, ok := s.streams[channelName]
+	if ok {
```
Go question - can channels be closed? Does this need to check whether the channel is closed prior to returning it? Maybe creating another if it's closed?

louisheal replied:
Channels can be closed, but there isn't an obvious way to check if they are
Since we're managing the channels ourselves within this struct, we know that we're not closing them, so they should never be closed

louisheal committed:
refresh access token only when inavlidated

louisheal committed:
remove viewers endpoint

louisheal committed:
rename user id to Twitch id

louisheal committed:
create get channels viewers method on channel service
write tests for get channels viewers
returned channels viewers as join server sent events

louisheal committed:
implemented get channels viewers method on spy
